accepted/future-releases/nnbd/feature-specification.md

```diff
@@ -263,6 +268,167 @@ The **flatten** function is modified as follows:
+Then the return statement $s$ completes returning $o$
+(\ref{statementCompletion}).
+
+\commentary{%
+The case where the evaluation of $e$ throws is covered by the general rule
+which propagates the throwing completion from $e$ to $s$ to the function body.%
+}
+
+When $f$ is an asynchronous non-generator with future value type $T_v$
+(\ref{functions}),
+the return statement $s$ is desugared at compile time.
+In the definition, the first item whose requirement is satisfied is used,
+and $\Gamma$ denotes the typing environment of $s$:
+
+\begin{itemize}
+\item% 1
+  When $S$ is \DYNAMIC:
+  \code{\VAR{} r = $e$; \RETURN{} r \IS{} Future<$T_v$> ? \AWAIT{} r : (r \AS{} $T_v$);}
+\item% 2
+  When $\flatten{S} \not= S$ and \SubtypeStd{\flatten{S}}{T_v}:
+  % flatten(S) != S implies that `Future` or `FutureOr` occurs.
+  \begin{itemize}
+  \item% 2a
+    When \SubtypeStd{S}{\code{Future<$T_v$>}} or
+    $S$ is of the form \code{$S_1$?} such that
+    $S_1$ implements \code{Future<$U$>} and \SubtypeStd{\code{$U$}?}{T_v}:
+    \code{\RETURN{} \AWAIT{} $e$;}
+  \item% 2b
+    When $S$ is of the form \code{FutureOr<$U$>} where \SubtypeStd{U}{T_v},
+    or $S$ is of the form \code{FutureOr<$U$>?}
+    where \SubtypeStd{\code{$U$?}}{T_v}:
+    \code{\VAR{} r = $e$; \RETURN{} r is Future<$T_v$> ? \AWAIT{} r : r;}
+  \end{itemize}
+\item% 3
+  When \SubtypeStd{S}{T_v}: \code{\RETURN{} $e$;}
+\end{itemize}
+
+\commentary{%
+Note that one of the above cases apply, or a compile-time error has occurred.%
+% It is required at compile-time that S is assignable to T_v, or that
+% flatten(S) <: T_v.
+%
+% If S is assignable to T_v then at least item 1 or 3 applies.
+%
+% Otherwise flatten(S) <: T_v. If flatten(S) == S, then S <: T_v, which is the
+% previous case (and 1 or 3 applies).
+%
+% So S != flatten(S) <: T_v. In this situation there will be an item
+% under item 2 that applies: S is of the form FutureOr<S1> where S1 <: T_v; or
+% S is of the form FutureOr<S1>? where S1? <: T_v; or S implements Future<S1>
+% and flatten(S) == S1; or S is S1? such that S1 implements Future<S2>, and
+% S2? <: T_v. In each case, subitem 2a or 3b applies.
+}
+
+% The new semantics preserves the pre-null-safety semantics with the following
+% exceptions:
+%
+% The old semantics tests for `is Future<flatten(T)>` where `T` is the declared
+% return type; the new semantics tests for `is Future<T_v>` where `T_v` is the
+% future value type. They differ only for null-safety specific types of the
+% forms `Future<U>?`, `FutureOr<U>?`, or `Object`. So the use of 'future value
+% type' causes no difference in behavior.
+%
+% item 1: Identical behavior, we always await a `Future<T_v>`, never await any
+% other type of object (e.g., we don't await a `Future<dynamic>`), and always
+% guard an object which is not awaited with a dynamic type check.
+%
+% item 2: New semantics is identical to old semantics.
+%
+% item 3: Breaking change when the value of $e$ has type `Future<T_v> & T_v`,
```

Comment:
I don't think this is right.  I think this is a much bigger change than implied here.  In particular, I believe that this proposal changes the following program from printing `int` to printing `Future<int>`:

```
Object foo() async {
  Object f = Future<Object>.value(3);
  return f;
}
void main() async {
  var f = foo();
  print((await f).runtimeType);
}
```

Reply:
Yes, I gave the following concrete example in an earlier comment:

```dart
Future<Object?> f() async => Future.value(1) as Object?;
```

This will complete the returned future with `1` with the old semantics and it will complete it with the future with the new semantics.

Reply:
This issue was eliminated by going back to rules that are much more like the current rules.